ompi/mpi/fortran/base/gen-mpi-mangling.pl

```diff
@@ -64,33 +64,33 @@
 $fortran->{unweighted} = {
     c_type => "int *",
     c_name => "mpi_fortran_unweighted",
-    f_type => "integer",
+    f_type => "integer, dimension(2)",
     f_name => "MPI_UNWEIGHTED",
 };
 $fortran->{weights_empty} = {
     c_type => "int *",
     c_name => "mpi_fortran_weights_empty",
-    f_type => "integer",
+    f_type => "integer, dimension(2)",
     f_name => "MPI_WEIGHTS_EMPTY",
 };
 
 $fortran->{argv_null} = {
     c_type => "char *",
     c_name => "mpi_fortran_argv_null",
-    f_type => "character, dimension(1)",
+    f_type => "character, dimension(8)",
```

Comment:
Do we need checks here to adjust for 32 or 64 bits?  I.e., do we know that these sizes will be the same as those in `libmpi.*`?

I ask because I don't fully grok the ld issue, nor the workaround -- i.e., why making the symbols bigger actually makes it better.

Reply:
Fortran object files and `libmpi.so` are used to link `libmpi_usempif08.so`. We need to make sure st_size (1 before this change) of common symbols in Fortran object files are not smaller than st_size in `libmpi.so` (4 on 32-bit platforms). Increasing it to 8 does not matter (just like st_size=1 does not matter before this change).

Reply:
Forgive me: I see _what_ you are doing here, I still don't understand _why_, though.  I know enough about linkers to know that every time I think I understand them, I find out that I don't know jack about linkers.  😦 

Let's discuss one of these common symbol constants in detail (the others are all generally the same):

1. In C, `mpi_fortran_argv_null` is declared in `ompi/include/mpif-c-constants.h` as `char *mpi_fortran_argv_null_`.  In Fortran -- originally from this Perl script -- `MPI_ARGV_NULL` is declared as `character, dimension(1)`.   It's an array of size 1 so that it would match the dereferenced C type (i.e., `(char*)`).  If we bump the Fortran array to size 8, do we need to make the C declaration be `char mpi_fortran_argv_null_[8]`?  Or is that naieve, and it was never about matching the size of the _dereferenced_ item -- since Fortran passes by reference, it was always about matching the size of the _pointer to_ the item?  That would seem weird, though...
1. I'm unclear on why the symbol in Fortran has to be greater than or equal to the size of the C variable.  Why doesn't it have to be the _same_ size?

Also, what will this do in versions of ld that do not have this bug?

Finally, what will this do to our ABI?  We unfortunately have strong ABI guarantees within a single release series (e.g., within the v4.0.x series).

Reply:
> If we bump the Fortran array to size 8, do we need to make the C declaration be char mpi_fortran_argv_null_[8]?

MPI 3.1 defines types of these C/Fortran constants, but I am very unclear about how these constants are used. It does appear weird to me that in C, it is a pointer while in Fortran, it is a character array. Due to 2., we have to make st_size of common symbols consistent.

> 2. I'm unclear on why the symbol in Fortran has to be greater than or equal to the size of the C variable. Why doesn't it have to be the same size?

`ompi/mpi/fortran/base/gen-mpi-mangling.pl` does not know the wordsize of the target platform. Ideally the same sizes are the best.

> Also, what will this do in versions of ld that do not have this bug?

The exact behavior is described in https://sourceware.org/bugzilla/show_bug.cgi?id=25236

1) The definition os `mpi_fortran_argvs_null_` will be provided by libmpi.so. It seems I have a debate with Alan Modra in that bug as whether resolving the symbol to libmpi.so makes sense. Alan insists that GNU ld's behavior is correct. Though it is at odds with other symbols where definitions in object files always win.
2) `mpi_fortran_argvs_null_` gets an incorrect version `VER_NDX_LOCAL`.

> Finally, what will this do to our ABI? We unfortunately have strong ABI guarantees within a single release series (e.g., within the v4.0.x series).

`st_size` is the only affected field. It may affect an ELF hack: copy relocations, which only appears in non-pic case. Fortunately, we probably don't have to consider copy relocations, because version symbols were not correctly assigned to these symbols, so it never worked.

Reply:
Sorry for the delay in replying / thanks for the reminder ping (your last post was right before I traveled for work, and that always makes a disaster of my INBOX ☹️).

> MPI 3.1 defines types of these C/Fortran constants, but I am very unclear about how these constants are used. 

Let me take a step back and explain what we're trying to do here.

1. MPI-3.1 defines these constants that can be used as sentinel values, passed in via parameters to MPI functions.  E.g., a Fortran app can call an MPI subroutine and pass `MPI_ARGV_NULL` as one of the parameters.  This sentinel will then trigger some specific behavior in that MPI subroutine.
1. Open MPI's base-level Fortran bindings for MPI subroutines are actually written in C.  Meaning: we do some testing in `configure` to figure out which of the 4 common symbol nomenclatures are used for Fortran subroutines, and use some C preprocessor magic to make our C implementations of the Fortran bindings adhere to that nomenclature (sometimes through weak symbols).
    * The implication here is that Fortran user codes are directly calling C -- and not via ISO(C) binding.  We do this because not all Fortran compilers that we have to support include support for ISO C binding.  And even if they did, there's some parameters that do not translate to ISO C binding (specifically: Fortran `LOGICAL`).  Ugh!
    * As such, all subroutine parameters are received in that C function as references -- i.e., pointers to that actual values.
    * Hence, if you pass a Fortran `INTEGER` parameter, we receive it in C as `(int*)` (assuming Fortran `INTEGER` and C `int` are the same size).
1. Therefore, for sentinels like `MPI_ARGV_NULL`, the Fortran version of the symbol must be the correct type to pass through the strict Fortran interface.
    * All we care about in C is testing if a given parameter matches a sentinel symbol (i.e., to know if `MPI_ARG_NULL` was passed or not).

Is there a better way to do this?

The limiting factor here is that the Fortran bindings are written in C, so we need to be able to compare these parameters for equality with sentinel values in C.  Some of these MPI routines are _highly_ performance critical; making an additional call out to an actual Fortran routine to do the comparison is... undesirable (i.e., it would add overhead in a performance-sensitive section of code).

> It does appear weird to me that in C, it is a pointer while in Fortran, it is a character array. Due to 2., we have to make st_size of common symbols consistent.

Ok.  If the real issue is that somehow it has been working for years even though our sizes have been inconsistent between C and Fortran but something broke -- at least in one released version of the linker -- such that our sloppiness doesn't work, but it _does_ work if the sizes are equal / we're not sloppy, then I would agree with you that this strategy is an improvement and we should do it.

> `ompi/mpi/fortran/base/gen-mpi-mangling.pl` does not know the wordsize of the target platform. Ideally the same sizes are the best.

FWIW: we can fix that (i.e., to make the script know the wordsize of the target).

> > Also, what will this do in versions of ld that do not have this bug?
> 
> The exact behavior is described in https://sourceware.org/bugzilla/show_bug.cgi?id=25236
> 
> 1. The definition os `mpi_fortran_argvs_null_` will be provided by libmpi.so. It seems I have a debate with Alan Modra in that bug as whether resolving the symbol to libmpi.so makes sense. Alan insists that GNU ld's behavior is correct. Though it is at odds with other symbols where definitions in object files always win.
> 2. `mpi_fortran_argvs_null_` gets an incorrect version `VER_NDX_LOCAL`.

I wonder if we're approaching this wrong.  We shouldn't be trying to change the _Fortran_ type of `MPI_ARGV_NULL` -- we should be changing the _C_ type of the symbol that we're using to compare.

Specifically: Fortran is going to pass the address of the Fortran symbol `MPI_ARGV_NULL` to the MPI function.  We have a corresponding C symbol, and we basically do this comparison in C to see if that sentinel value was passed:

```c
extern char * mpi_fortran_argv_null_;
#define OMPI_IS_FORTRAN_ARGV_NULL(addr) \
       (addr == (void*) &mpi_fortran_argv_null_)

// This is the C function implementing the Fortran binding for MPI_COMM_SPAWN
void ompi_comm_spawn_f(char *command, char *argv, MPI_Fint *maxprocs,
                      MPI_Fint *info, MPI_Fint *root, MPI_Fint *comm,
                      MPI_Fint *intercomm, MPI_Fint *array_of_errcodes,
                      MPI_Fint *ierr, int cmd_len, int string_len)
{
    // ...
    if (OMPI_IS_FORTRAN_ARGV_NULL(argv)) {
        // ...do ARGV_NULL things...
    }
    // ...
}
```

With this example, I guess I'm still confused as to why the Fortran `MPI_ARGV_NULL` (which is of size 1) is not equal in size to the C `mpi_fortran_argv_null_` (which is also of size 1).  I'm also confused as to why this matters, because all we're doing is comparing the address of the parameter that is passed in (i.e., `argv`) to `&mpi_fortran_argv_null_` -- the size of the dereferenced symbols shouldn't matter...?

What am I missing?